libbeat/common/transport/tlscommon/verify.go

```diff
@@ -98,3 +92,64 @@ func verifyCertificateExceptServerName(
 	chains, err := headCert.Verify(opts)
 	return certs, chains, err
 }
+
+func verifyCertificateWithLegacyCommonName(
+	rawCerts [][]byte,
+	c *TLSConfig,
+) ([]*x509.Certificate, [][]*x509.Certificate, error) {
+	certs, err := overwriteSANWithCommonName(rawCerts)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to overwrite SAN with CommonName")
+	}
+
+	var t time.Time
+	if c.time != nil {
+		t = c.time()
+	} else {
+		t = time.Now()
+	}
+
+	headCert := certs[0]
+
+	opts := x509.VerifyOptions{
+		DNSName:       headCert.DNSNames[0],
+		Roots:         c.RootCAs,
+		CurrentTime:   t,
+		Intermediates: x509.NewCertPool(),
+	}
+
+	for _, cert := range certs[1:] {
+		opts.Intermediates.AddCert(cert)
+	}
+
+	// defer to the default verification performed
+	chains, err := headCert.Verify(opts)
+	return certs, chains, err
+}
+
+// overwriteSANWithCommonName adds the Common Name to DNSNames, if the list is empty.
+// It is a workaround to address the changes introduced in Golang 1.15: https://golang.org/doc/go1.15#commonname.
+func overwriteSANWithCommonName(
+	rawCerts [][]byte,
+) ([]*x509.Certificate, error) {
+	// this is where we're a bit suboptimal, as we have to re-parse the certificates that have been presented
+	// during the handshake.
+	// the verification code here is taken from verifyServerCertificate in crypto/tls/handshake_client.go:824
+	certs := make([]*x509.Certificate, len(rawCerts))
+	for i, asn1Data := range rawCerts {
+		cert, err := x509.ParseCertificate(asn1Data)
+		if err != nil {
+			return nil, errors.Wrap(err, "tls: failed to parse certificate from server")
```

Comment:
better use `fmt.Errorf("tls: failed to parse certificate from server: %w", err)`